framework/data_types/ndarray.h

```diff
@@ -48,341 +50,147 @@ class NDArray
-    size_ = 1;
-    for (size_t i = 0; i < rank_; ++i)
-    {
-      assert(dimensions_[i] > 0);
-      size_ *= dimensions_[i];
-      strides_[i] = 1;
-      for (size_t j = i + 1; j < rank_; ++j)
-        strides_[i] *= dimensions_[j];
-    }
-
-    base_ = new T[size_];
-    for (size_t i = 0; i < size_; ++i)
-      base_[i] = value;
+    set_dimensions(dims);
+    value_initialize(value);
   }
 
-  /**
-   * Creates an empty array.
-   * \throw std::bad_alloc if memory allocation fails.
-   *
-   * This constructor creates an empty array and initializes the reference
-   * count to one.
-   */
-  NDArray() : rank_(0), dimensions_(), strides_(), size_(0), base_(nullptr) {}
-
-  /**
-   * Copy construct from another array.
-   *
-   * \param other The array to copy.
-   */
-  NDArray(NDArray<T> const& other)
+  /// Copy constructor
+  NDArray(const NDArray<T>& other)
     : rank_(other.rank_),
-      dimensions_(other.rank_),
-      strides_(other.rank_),
       size_(other.size_),
-      base_(nullptr)
+      storage_(std::make_unique<T[]>(other.size_)),
+      dimensions_(other.dimensions_),
+      strides_(other.strides_)
   {
-    base_ = new T[size_];
-
-    for (size_t i = 0; i < rank_; ++i)
-    {
-      dimensions_[i] = other.dimensions_[i];
-      strides_[i] = other.strides_[i];
-    }
-
-    for (size_t i = 0; i < size_; ++i)
-      base_[i] = other.base_[i];
-  }
-
-  /**
-   * Assign from another array.
-   *
-   * \param other The array to copy.
-   */
-  NDArray<T>& operator=(NDArray<T> const& other)
-  {
-    NDArray<T>(other).swap(*this);
-    return *this;
+    std::copy(other.storage_.get(), other.storage_.get() + size_, storage_.get());
   }
 
   /// Move constructor
   NDArray(NDArray<T>&& other) noexcept
-    : rank_(std::move(other.rank_)),
+    : rank_(other.rank_),
+      size_(other.size_),
+      storage_(std::move(other.storage_)),
       dimensions_(std::move(other.dimensions_)),
-      strides_(std::move(other.strides_)),
-      size_(std::move(other.size_)),
-      base_(std::move(other.base_))
-  {
-  }
-
-  /// Deleted move assignment operator
-  NDArray<T>& operator=(NDArray<T>&&) = delete;
-
-  /// Sets a value to all the items in the array.
-  void set(T value)
+      strides_(std::move(other.strides_))
   {
-    for (size_t i = 0; i < size_; ++i)
-      base_[i] = value;
+    other.rank_ = 0;
+    other.size_ = 0;
   }
 
-  /**
-   * Swap the contents of this array with another array.
-   *
-   * \param other The array to swap with.
-   */
-  void swap(NDArray<T>& other)
+  /// Copy assignment operator
+  NDArray<T>& operator=(const NDArray<T>& other)
   {
-    std::swap(rank_, other.rank_);
-    std::swap(dimensions_, other.dimensions_);
-    std::swap(strides_, other.strides_);
-    std::swap(size_, other.size_);
-    std::swap(base_, other.base_);
+    if (this != &other)
+    {
+      NDArray<T> temp(other); // Copy constructor, uses swap to reduce redundancy
+      this->swap(temp);
+    }
```

Comment:
```suggestion
    if (this != &other)
    {
      rank_ = other.rank_;
      size_ = other.size_;
      dimensions_ = other.dimensions_;
      strides_ = other.strides_;
      storage_ = other.storage_
      std::copy(other.storage_.get(), other.storage_.get() + size_, storage_.get());
    }
```
Why allocate a new NDArray here instead of implementing this more like the move assignment operator?

Reply:
It's much faster if we swap than doing the `std::copy`.

Reply:
Let me look at this. You're right that we call the copy constructor anyway, so, from a performance perspective, it's probably the same regardless.

Reply:
I think if you do just `storage_ = other.storage_;`, then you leak the memory associated with `storage_`. This is why I think the original code uses `swap`. The old pointer gets assigned into `other` and when it goes out of scope, the memory gets deallocated...

Reply:
But, this is not true anymore, since `storage_` is now `unique_ptr`... Just an explanation why `swap` was there...

Reply:
Yeah, I suspect that the `swap` predates my use of `unique_ptr`.